### PR TITLE
Handle resources pagination

### PR DESCRIPTION
**Pull request by daniloisr**
Use `collection.next` to fetch the next page using jsonapi pagination links.

### CODE REVIEW

File: lib/munson/collection.rb [hunk 1]
```diff
@@ -18,5 +19,11 @@ def initialize(collection=[], errors: nil, meta: nil, jsonapi: nil, links: nil)
     def each(&block)
       @collection.each(&block)
     end
+
+    def next
+      return self.class.new if !links || !links[:next]
+
+      @collection.first.class.from_url(links[:next])
```
Comment (thiagoa) [inline lib/munson/collection.rb]: I see some problems with this code, so what do you think about the following approach:

```ruby
def next
  return self.class.new if !links || !links[:next]
  
  response = Munson.default_connection.get(path: links[:next])
  ResponseMapper.new(response.body).collection
end
```

This would allow us to get rid of the delegation and the `from_url` method on `Client`. The current code:

- Is reaching into an instance and delegating a long way just to get to a client connection and use it in a generic way. 
- Is relying on the presence of at least one collection item. What if the response comes in defective?
- From what I see, your code would work the same regardless of the first item's resource class: this is a great clue that you can access a global resource instead of using a client that is configured specifically for an instance (and using it in a generic way).

File: lib/munson/query.rb [hunk 2]
```diff
@@ -35,6 +35,13 @@ def find(id)
       end
     end
 
+    def from_url(url)
+      raise Munson::ClientNotSet, "Client was not set. Query#new(client)" unless @client
```
Comment (thiagoa) [inline lib/munson/query.rb]: A refactor idea: we could create a `client` method:

```ruby
def client
  @client || raise ........
end
```

That would eliminate some repetition, but this PR is probably not appropriate to do it.

File: lib/munson/resource.rb [hunk 3]
```diff
@@ -176,13 +177,5 @@ def fields(*args)
       hash_fields[type] = args if args.any?
       munson.fields(hash_fields)
     end
-
-    [:include, :sort, :filter, :fetch, :find, :page].each do |method|
```
Comment (thiagoa) [inline lib/munson/resource.rb]: I wonder what was the intention of this code. I suspect the author's intent was to make the delegation faster, but I'm not sure.

File: spec/munson/collection_spec.rb [hunk 4]
```diff
@@ -4,4 +4,15 @@
   specify { expect(subject).to respond_to(:meta) }
   specify { expect(subject).to respond_to(:jsonapi) }
   specify { expect(subject).to respond_to(:links) }
+
+  describe '#next' do
+    it 'fetches next collections' do
+      stub_api_request(:articles_with_pagination_1)
+      stub_api_request(:articles_with_pagination_2)
+
+      articles = Article.page(number: 1, size: 1).fetch.next
```
Comment (thiagoa) [inline spec/munson/collection_spec.rb]: What do you think of splitting this into two variables to make it easier to read:

```ruby
page_1_articles = Article.page(number: 1, size: 1).fetch

page_2_articles = page_1_articles.next

# Expectations... 
```

File: spec/munson/collection_spec.rb [hunk 5]
```diff
@@ -4,4 +4,15 @@
   specify { expect(subject).to respond_to(:meta) }
   specify { expect(subject).to respond_to(:jsonapi) }
   specify { expect(subject).to respond_to(:links) }
+
+  describe '#next' do
```
Comment (thiagoa) [inline spec/munson/collection_spec.rb]: Maybe include a test case to exercise a response with empty data and a "next" link? Not sure if it would be worth.

File: README.md [hunk 6]
```diff
@@ -116,6 +116,15 @@ query.to_params
 query.fetch #=> Munson::Collection<Product,Product>
 ```
 
+#### Paging
+
+```ruby
+collection = Product.page(number: 1, limit: 20).fetch
+collection.size #=> 20
+
+collection.next #=> fetches next page
```
Comment (thiagoa) [inline README.md]: What do you think of mentioning that this feature relies on the presence of a `links[:next]` attribute?

File: lib/munson/collection.rb (hunk 1 above)
Comment (daniloisr) [inline lib/munson/collection.rb]: Good points, but I'll use it partially, because we can't rely on `Munson.default_connection`, as each resource can have its own munson connection and add things like headers, ...

So I'll just drop the method `from_url` and use the `.get` direct here.

> Is reaching into an instance and delegating a long way

there is a lot of delegation in munson, and I particularly dislike it. I did a spike to reduce the number of delegations but it will require a big effort, which didn't fit on my milestone.

> Is relying on the presence of at least one collection item. What if the response comes in defective?

Good point. I'll handle the case where I can't get the first collection item. And about the defective response, this is handled in the mapper, which will not find a resource class and throw an error.

File: lib/munson/query.rb (hunk 2 above)
Comment (daniloisr) [inline lib/munson/query.rb]: Yes, you suggestion will reduce a lot of `if`s that we have now, but let's do it in a new PR

File: lib/munson/resource.rb (hunk 3 above)
Comment (daniloisr) [inline lib/munson/resource.rb]: > I suspect the author's intent was to make the delegation faster, but I'm not sure.

Do you know how much fast it will be using `class_eval`? I rather think that the author didn't knew that was possible to delegate to class variables.

Comment (thiagoa) [inline lib/munson/resource.rb]: That may be it.

File: lib/munson/collection.rb (hunk 1 above)
Comment (thiagoa) [inline lib/munson/collection.rb]: Suppose we query the articles API but for some reason, we get a response similar to the following:

```ruby
{
  data: [
    { type: 'authors', ....... },
    { type: 'articles', ........ }
  ]
}
```

I'm supposing Munson allows this kind of flexibility (correct me if I'm wrong), because each data item can have a distinct type (instead of the whole thing having a type).

In that case, your code would use an `Author` resource to query articles and possibly send wrong headers (in case `Author` defines distinct headers). I guess the safest alternative would be to "reverse engineer" the URL by extracting its path component:

```ruby
  return self.class.new if !links || !links[:next]

  type = URI(links[:next]).path[1..-1]
  klass = Munson.lookup_type(type) || fail 'Invalid URL'

  response = klass.munson.connection.get(path: links[:next])
  ResponseMapper.new(response.body).collection
```

That way we would make sure to use the right resource. What do you think?

Comment (daniloisr) [inline lib/munson/collection.rb]: I almost sure that this response is not supported by JSON-API specs: http://jsonapi.org

So we'll not face this kind of problem.

Comment (thiagoa) [inline lib/munson/collection.rb]: If it's not supported then we're good.

**Comment by prashantrajan** [comment]
LGTM. There's a conflict on the spec that needs to be resolved.
